### PR TITLE
satellite_data

### PR DESCRIPTION
**Pull request by wabinyai**
**_WHAT DOES THIS PR DO?_**

1. Query AirQo data
2. Pull satellite data from Google Earth engine
3. Merged the data with Airqo device
4. Saves the merged data

- [ ] Item 1
- [ ] Item 2

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
    - []
- GitHub issues
    - Closes #<enter GitHub issue number here>

**_HOW DO I TEST OUT THIS PR?_**

- [ ] Analytics: [Link to Analytics README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/analytics#readme)
- [ ] Auth Service: [Link to Auth Service README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/auth-service#readme)
- [ ] Calibrate: [Link to Calibrate README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/calibrate#readme)
- [ ] Data Proxy: [Link to Data Proxy README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/data-proxy#readme)
- [ ] Device Monitoring: [Link to Device Monitoring README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-monitoring#readme)
- [ ] Device Registry: [Link to Device Registry README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-registry#readme)
- [ ] Device Status: [Link to Device Status README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-status#readme)
- [ ] Device Uptime: [Link to Device Uptime README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-uptime#readme)
- [ ] Exceedances: [Link to Exceedances README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/exceedances#readme)
- [ ] Firebase: [Link to Firebase README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/firebase)
- [ ] GP Model: [Link to GP Model README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/gp-model#readme)
- [ ] Incentives: [Link to Incentives README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/incentives#readme)
- [ ] Kafka Connectors: [Link to Kafka Connectors](https://github.com/airqo-platform/AirQo-api/tree/staging/src/kafka-connectors)
- [ ] Locate: [Link to Locate README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/locate#readme)
- [ ] Metadata: [Link to Metadata README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/meta-data#readme)
- [ ] Predict: [Link to Predict README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/predict#readme)
- [ ] Spatial: [Link to Spatial](https://github.com/airqo-platform/AirQo-api/tree/staging/src/spatial)
- [ ] View: [Link to View README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/view#readme)
- [x] Workflows: [Link to Workflows README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/workflows#readme)

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [ ] Endpoint 1
- [ ] Endpoint 2

**_ARE THERE ANY RELATED PRs?_**

- [ ] Related PR 1
- [ ] Related PR 2

### CODE REVIEW

File: src/workflows/dags/pull__satellite_data.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    last_timestamp = None
+
+    while start_time <= end_time:
+        print("Querying BigQuery...")
+        data = data_handler.query_bigquery_batch(start_time=start_time, end_time=end_time, batch_size=batch_size)
+        if data.empty:
+            break
+        print("Querying BigQuery complete.")
+
+        print("Processing geolocation data...")
```
Comment (Mnoble-19) [inline src/workflows/dags/pull__satellite_data.py:36]: @wabinyai its not a good practice to leave print statements in production codebases. if you wish to view logs, you can use the python logger, however airflow also does this by default (check out one of our production dags while its running. 
```suggestion
  
```
[here](https://medium.com/red-buffer/python-production-level-coding-practices-4c39246e0233#:~:text=Logging%20Critical%20Failures%20and%20Intermediate%20Results)

File: src/workflows/airqo_etl_utils/satellite_data_utils.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    key_file=Config.GOOGLE_APPLICATION_CREDENTIALS,
+    email=Config.GOOGLE_APPLICATION_CREDENTIALS_EMAIL,
+)
+class DataHandler:
+    def __init__(self):
+        self.client = bigquery.Client()
+        self.data_path = None
+        ee.Initialize(credentials, project=Config.GOOGLE_CLOUD_PROJECT_ID)
+        self.mongo_client = MongoClient(Config.MONGODB_URI)
+        self.db = self.mongo_client[Config.MONGODB_DATABASE]
```
Comment (Mnoble-19) [inline src/workflows/airqo_etl_utils/satellite_data_utils.py:29]:
```suggestion
```
@wabinyai there is already a connection created at the bottom of `config.py` so you can imiport it and skip this. perhaps create a new collection only.

File: src/workflows/airqo_etl_utils/satellite_data_utils.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+            query_job = self.client.query(query)
+            df = query_job.to_dataframe() 
+            df = df.sort_values(by=['site_name', 'timestamp']).reset_index(drop=True)
+            df['date'] = [str(x.date()) for x in df.timestamp]
+            df['month'] = df.timestamp.dt.month
+            df['hour'] = df.timestamp.dt.hour
+            return df
+        except Exception as e:
+            print(f"Error querying BigQuery: {e}")
+            return None
```
Comment (Mnoble-19) [inline src/workflows/airqo_etl_utils/satellite_data_utils.py:71]: @wabinyai most of these bigquery related methods are in `bigquery_api.py` so you could move this there

File: src/workflows/airqo_etl_utils/satellite_data_utils.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                    self.collection.update_one(
+                        {'_id': existing_record['_id']},
+                        {'$set': new_data}
+                    )
+                else:
+                    # Insert new record
+                    self.collection.insert_one(record)
+            print("Data saved to MongoDB successfully.")
+        except Exception as e:
+            print(f"Error saving data to MongoDB: {e}")
```
Comment (Mnoble-19) [inline src/workflows/airqo_etl_utils/satellite_data_utils.py:228]: Hi @wabinyai mongodb has an `upsert` parameter that allows you to update existing ones, so you can leave out the if else, the code should look somethinig like this after. 
```suggestion
def save_to_mongodb(self, merged_df_):
    try:
        for record in merged_df_.to_dict(orient='records'):
            # Define a filter for identifying the unique document
            filter_ = {
                'site_id': record['site_id'],
                'date': record['date'],
                'hour': record['hour']
            }

            # We use $set to update fields or insert them if they don't exist
            update = {'$set': record}

            # Perform an upsert (update existing or insert new if none exists)
            self.collection.update_one(filter_, update, upsert=True)
    except Exception as e:
        print(f"Error saving data to MongoDB: {e}")

```

File: src/workflows/airqo_etl_utils/config.py [hunk 5]
```diff
@@ -205,6 +205,12 @@ class Config:
     TWITTER_BOT_ACCESS_TOKEN = os.getenv("TWITTER_BOT_ACCESS_TOKEN")
     TWITTER_BOT_ACCESS_TOKEN_SECRET = os.getenv("TWITTER_BOT_ACCESS_TOKEN_SECRET")
 
+    #Pull data from satellite
+    BIGQUERY_SITES = os.getenv("BIGQUERY_SITES")
+    GOOGLE_APPLICATION_CREDENTIALS_EMAIL = os.getenv("GOOGLE_APPLICATION_CREDENTIALS_EMAIL")
+    MONGODB_URI = os.getenv("MONGODB_URI")
+    MONGODB_DATABASE = os.getenv("MONGODB_DATABASE")
+    MONGODB_COLLECTION = os.getenv("MONGODB_COLLECTION")
```
Comment (Mnoble-19) [inline src/workflows/airqo_etl_utils/config.py:213]: hello @wabinyai, some of these secrets already exist e.g MONGOURI, so no need for new uris maybe DBs, amd  perhaps a new collection i know MONGO_COLLECTION already exists, so perhaps use SATELITE_MONGO_COLLECTION . Skim through the list of existing ones in this very config file
```suggestion
    BIGQUERY_SITES = os.getenv("BIGQUERY_SITES")
    GOOGLE_APPLICATION_CREDENTIALS_EMAIL = os.getenv("GOOGLE_APPLICATION_CREDENTIALS_EMAIL")
    MONGODB_URI = os.getenv("MONGODB_URI")
    MONGODB_DATABASE = os.getenv("MONGODB_DATABASE")
    SATELITE_DATA_COLLECTION = os.getenv("SATELITE_DATA_COLLECTION")
```

File: src/workflows/airqo_etl_utils/satellite_data_utils.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        for key in site_dfs.keys():
+          result_df = pd.concat([result_df, site_dfs[key]])
+        return result_df
+    
+    def extract_and_merge_data(self, df, result_df):
+      df1 = df[['site_id', 'site_name', 'pm2_5', 'pm10', 'date', 'site_latitude', 'site_longitude', 'country', 'month', 'hour']]
+      merged_df_ = df1.merge(result_df, how='right', on=['site_id', 'date', 'month', 'hour'])
+      merged_df_= merged_df_[~merged_df_['country'].isna()].reset_index(drop = True)
+      merged_df_ = merged_df_.drop(columns=[col for col in merged_df_.columns if '_datetime' in col])
+      merged_df_.to_csv('output.csv', index=False)
```
Comment (Mnoble-19) [inline src/workflows/airqo_etl_utils/satellite_data_utils.py:202]: Hi @wabinyai , why are you saving this CSV?
```suggestion

```

File: src/workflows/airqo_etl_utils/setup.py [hunk 7]
```diff
@@ -35,6 +35,9 @@
         "pymongo~=4.5.0",
         "optuna",
         "tweepy",
+        "ee",
+        "gunicorn"
```
Comment (Mnoble-19) [inline src/workflows/airqo_etl_utils/setup.py:39]: Why is this needed? @wabinyai

**Review by Mnoble-19 — CHANGES_REQUESTED** [review]

File: src/workflows/airqo_etl_utils/satellite_data_utils.py (hunk 6 above)
Comment (wabinyai) [reply src/workflows/airqo_etl_utils/satellite_data_utils.py:202]: It is for testing but not realy needed

File: src/workflows/dags/pull__satellite_data.py (hunk 1 above)
Comment (wabinyai) [reply src/workflows/dags/pull__satellite_data.py:36]: You are right, the print was for testing if the line was running fine. These will be removed before it is merged into staging/production
